Generator Expressions

- generator expressions are more memory efficient alternative to list comprehensions

In [1]:
large_num = 99999999
l_squares = [x**2 for x in range(large_num)]

In [2]:
import sys
sys.getsizeof(l_squares)

835128600

- the only difference in syntax between list comprehensions and generators is that generators use () brackets

In [3]:
g_squares = (x**2 for x in range(large_num))
sys.getsizeof(g_squares)

208

In [4]:
g_squares

<generator object <genexpr> at 0x7f250eeaf030>

- the size above is small (208) because generator expressions calculate values as they're requested in a similar way to range object
- to access the values of a generator expression, we can use the next function

In [5]:
next(g_squares)

0

- calling next expression generates the forward value each time

In [6]:
for x in g_squares:
    print(x)
    if x > 12:
        break

1
4
9
16


Chaining Generator Expression

In [8]:
evens = (x for x in range(0, 100, 2))
div_three = (y for y in evens if y%3 == 0)
div_three

<generator object <genexpr> at 0x7f250eeafb90>

In [9]:
next(div_three)

0

In [10]:
[x for x in div_three]

[6, 12, 18, 24, 30, 36, 42, 48, 54, 60, 66, 72, 78, 84, 90, 96]